### PR TITLE
[data, perf, ops] feat: option to pad packed input to a fixed shape for text-only models

### PR DESCRIPTION
**Pull request by Luosuu**
### What does this PR do?

Related issues: #401 #402 

VeOmni expects data to be feed with `rmpad_with_pos_ids` which would pack sequences into one batch and lead to different input length for every training iteration. This will consequently trigger recompile of the LigerRMSNorm backward triton kernel, blocking GPU execution (often observed in the first backward layer).

We introduce a new data collator that remove padding with pos_ids and then pad the packed input to a fixed shape for convenience of CUDAGraph & torch.compile()

### Test

- [x] new data tests
- [x] varlen FA output accuracy test: padded vs. non-padded 
- [x] e2e smoke tests
  - [x] runnable
  - [x] accuracy 
  - [ ] perf 

### API and Usage Example

> Demonstrate how the API changes if any, and provide usage example(s) if possible.

```yaml
train:
  pad_packed_input: true
  pad_packed_to_length: 4096 # if not set this would be micro_batch_size * max_seq_len
```

### Design & Code Changes

Similar to SP, we directly pass the padded input into varlen FA kernel since we are providing `cu_seqlens`.
We add a test to verify that varlen FA can produce exact same outputs between padded and non-padded inputs with the same correct `cu_seqlens`

Effects: 
1. all the modeling are not touched
4. RMSNorm and MLP see the same shapes at the cost of additional compute for padded token

### CODE REVIEW

File: veomni/data/data_collator.py [hunk 1]
```diff
@@ -188,6 +189,46 @@ def __call__(self, features: Sequence[Dict[str, "torch.Tensor"]]) -> Dict[str, "
...
+            pad = torch.full(pad_shape, pad_value, dtype=tensor.dtype, device=tensor.device)
+            return torch.cat([tensor, pad], dim=-1)
+
+        batch["input_ids"] = pad_last_dim(batch["input_ids"], self.pad_token_id)
+        if "attention_mask" in batch:
+            batch["attention_mask"] = pad_last_dim(batch["attention_mask"], self.attention_mask_pad_value)
+        if "labels" in batch:
+            batch["labels"] = pad_last_dim(batch["labels"], IGNORE_INDEX)
+        if "position_ids" in batch:
+            batch["position_ids"] = pad_last_dim(batch["position_ids"], self.position_id_pad_value)
```
Comment (piyifan123) [inline veomni/data/data_collator.py]: do we need to care about cu_seqlens for padded tokens?

File: veomni/data/data_loader.py [hunk 2]
```diff
@@ -71,6 +72,8 @@ def build_native_dataloader(
     bsz_warmup_ratio: float = 0.02,
     bsz_warmup_init_mbtoken: int = 200,
     dyn_bsz: bool = True,
+    pad_packed_to_length: Optional[int] = None,
+    pad_packed_token_id: int = 0,
```
Comment (Coach257) [inline veomni/data/data_loader.py]: Do we really need this arg? input id always 0; labels always -100; position ids always 0; attention mask always 1.

Comment (Luosuu) [reply veomni/data/data_loader.py]: sure. thanks for the suggestion

File: veomni/data/data_collator.py (hunk 1 above)
Comment (Luosuu) [reply veomni/data/data_collator.py]: no, please see `tests/ops/test_flash_attn_varlen_padding.py`

File: docs/usage/rmpad_with_pos_ids_and_dyn_bsz.md [hunk 3]
```diff
@@ -73,3 +73,14 @@ Example fixed batch size = 2 samples:
 * Batch 2: S3 + S4
 * input_ids = [G H I J]
 * position_ids = [0 1 2 0]
+
+## Padding packed inputs (pad_packed_input)
+
+When `rmpad_with_pos_ids` is enabled, `pad_packed_input` can pad the packed sequence to a fixed length
+(`pad_packed_to_length`). This is useful to avoid uneven lengths that can trigger kernel recompilation.
+
```
Comment (piyifan123) [inline docs/usage/rmpad_with_pos_ids_and_dyn_bsz.md:82]: can we add an example here? Or link to the original github issue and pasted the before/after comparison trace there.

File: veomni/data/data_collator.py [hunk 4]
```diff
@@ -188,6 +189,53 @@ def __call__(self, features: Sequence[Dict[str, "torch.Tensor"]]) -> Dict[str, "
...
+        if pad_len == 0:
+            return batch
+
+        def pad_last_dim(tensor: "torch.Tensor", pad_value: int) -> "torch.Tensor":
+            pad_shape = list(tensor.shape)
+            pad_shape[-1] = pad_len
+            pad = torch.full(pad_shape, pad_value, dtype=tensor.dtype, device=tensor.device)
+            return torch.cat([tensor, pad], dim=-1)
+
+        batch["input_ids"] = pad_last_dim(batch["input_ids"], 0)
```
Comment (piyifan123) [inline veomni/data/data_collator.py:226]: should we follow the similar logic in L118 to pad?

File: veomni/arguments/arguments_types.py [hunk 5]
```diff
@@ -750,7 +758,22 @@ class VeOmniArguments:
     train: TrainingArguments = field(default_factory=TrainingArguments)
 
     def __post_init__(self):
-        pass
+        if self.train.pad_packed_input:
```
Comment (piyifan123) [inline veomni/arguments/arguments_types.py:761]: should we just assert that rmpad_with_pos_id must be true to have pad_packed_input=true?

File: docs/usage/rmpad_with_pos_ids_and_dyn_bsz.md (hunk 3 above)
Comment (Luosuu) [reply docs/usage/rmpad_with_pos_ids_and_dyn_bsz.md]: not sure how to share a pytorch trace with links. I will link the [#402](https://github.com/ByteDance-Seed/VeOmni/issues/402) here.

File: veomni/data/data_collator.py (hunk 4 above)
Comment (Luosuu) [reply veomni/data/data_collator.py:226]: I find that we cannot use pad_sequence as it does not accept a given length
